diff -ur popi-0.9/rlranklib.py popi-0.9-rpy2/rlranklib.py
--- popi-0.9/rlranklib.py	2011-08-17 15:30:31.000000000 +0200
+++ popi-0.9-rpy2/rlranklib.py	2011-09-16 05:28:16.547000068 +0200
@@ -5,10 +5,13 @@
 # $Header: /data/cvsroot/plportest/anadata/rlranklib.py,v 1.11.2.1 2007/11/26 08:46:29 lgh Exp $
 
 import random
-from rpy import *
+import rpy2.rpy_classic
+from rpy2.rpy_classic import *
 from math import sqrt, log
 from sets import Set
 
+rpy2.rpy_classic.set_default_mode(rpy2.rpy_classic.BASIC_CONVERSION)
+
 safe_eff = [None,1.365,1.208,1.116,1.110,1.104,1.064,1.064]
 
 def calc_rlrank(lrs, opt_ranktype):
@@ -66,7 +69,7 @@
         riag = [ rlranks[i] for i in range(0, len(rlranks)) if partition[i] == igrp ]    # rlranks in a group
         if len(riag) == 1:
             continue
-        threshold = r.qtukey(alpha, len(riag), 1000)*sqrt(1.0/12/nb)*len(riag)/len(rlranks)
+        threshold = r.qtukey(alpha, len(riag), 1000)[0]*sqrt(1.0/12/nb)*len(riag)/len(rlranks)
         range_    = max(riag) - min(riag)
         if verbose:
             print "   ", igrp, len(riag), min(riag), max(riag), range_, "%.3f" % threshold,
@@ -99,7 +102,7 @@
         return True
     
     # t = r.qtukey(alpha, len(arrs), 1000)*sqrt(1.0/12/nb)*(len(arrs)+1)/ksum
-    t = r.qtukey(alpha, km, 1000) * sqrt((km*km-1)/12.0/nb) / ksum
+    t = r.qtukey(alpha, km, 1000)[0] * sqrt((km*km-1)/12.0/nb) / ksum
     if km <= len(safe_eff):
         t = t * safe_eff[km-1]
     if max(arrs) - min(arrs) > t:
Binary files popi-0.9/rlranklib.pyc and popi-0.9-rpy2/rlranklib.pyc differ
